File: AOC_PY/days/day04.py
import os
import re

INPUT_PATH = os.path.join(os.path.dirname(__file__), '../../Inputs', 'day04.txt')


def validate_field(name: str, value: str):
    if name == 'byr':
        return re.match(r'^(19[2-9][0-9]|200[0-2])$', value) is not None
    elif name == 'iyr':
        return re.match(r'^(201[0-9]|2020)$', value) is not None
    elif name == 'eyr':
        return re.match(r'^(202[0-9]|2030)$', value) is not None
    elif name == 'hgt':
        return re.match(r'^(1[5-8][0-9]cm|19[0-3]cm|59in|6[0-9]in|7[0-6]in)$', value) is not None
    elif name == 'hcl':
        return re.match(r'^(#[0-9a-f]{6})$', value) is not None
    elif name == 'ecl':
        return re.match(r'^(amb|blu|brn|gry|grn|hzl|oth)$', value) is not None
    elif name == 'pid':
        return re.match(r'^(\d{9})$', value) is not None
    return True


class Day04:

    def __init__(self):
        self.passports = []
        with open(INPUT_PATH, 'r') as infile:
            lines = infile.read()
        for line in lines.split('\n\n'):
            self.passports.append(re.split(r'\n| |:', line.strip()))

    def solve1(self):
        solution = 0
        for passport in self.passports:
            fields = {'byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid'}
            for i in range(0, len(passport), 2):
                fields.discard(passport[i])
            if len(fields) == 0:
                solution += 1
        return solution

    def solve2(self):
        solution = 0
        for passport in self.passports:
            valid, fields = True, {'byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid'}
            for i in range(0, len(passport), 2):
                if validate_field(passport[i], passport[i + 1]):
                    fields.discard(passport[i])
                else:
                    valid = False
                    break
            if valid and len(fields) == 0:
                solution += 1
        return solution


def main():
    x = Day04()
    print(x.solve1())
    print(x.solve2())


if __name__ == '__main__':
    main()

File: AOC_CS/Days/Day04.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;

namespace AOC_CS.Days
{
    class Day04
    {
        const string INPUT_FILE = "Inputs\\day04.txt";

        private string[] passports;

        public Day04()
        {
            passports = null;
        }

        private void ReadInput()
        {
            if (passports != null) return;

            var lines = File.ReadAllText(INPUT_FILE);
            passports = lines.Split("\n\n");
        }

        public int Solve1()
        {
            ReadInput();

            var solution = 0;
            foreach (var pass in passports)
            {
                var requiredFields = new SortedSet<string> { "byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid" };
                var tokens = pass.Split(new[] { '\n', ' ', ':' }, StringSplitOptions.RemoveEmptyEntries);
                for (var i = 0; i < tokens.Length; i += 2)
                {
                    requiredFields.Remove(tokens[i]);
                }
                if (requiredFields.Count == 0) ++solution;
            }
            return solution;
        }

        private bool ValidateField(string field, string value)
        {
            switch (field)
            {
                case "byr":
                    return Regex.IsMatch(value, @"^(19[2-9][0-9]|200[0-2])$");
                case "iyr":
                    return Regex.IsMatch(value, @"^(201[0-9]|2020)$");
                case "eyr":
                    return Regex.IsMatch(value, @"^(202[0-9]|2030)$");
                case "hgt":
                    return Regex.IsMatch(value, @"^(1[5-8][0-9]cm|19[0-3]cm|59in|6[0-9]in|7[0-6]in)$");
                case "hcl":
                    return Regex.IsMatch(value, @"^(\#[0-9a-f]{6})$");
                case "ecl":
                    return Regex.IsMatch(value, @"^(amb|blu|brn|gry|grn|hzl|oth)$");
                case "pid":
                    return Regex.IsMatch(value, @"^([0-9]{9})$");
                case "cid":
                    return true;
                default:
                    return false;
            }
        }

        public int Solve2()
        {
            ReadInput();

            var solution = 0;
            foreach (var pass in passports)
            {
                var valid = true;
                var requiredFields = new SortedSet<string> { "byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid" };
                var tokens = pass.Split(new[] { '\n', ' ', ':' }, StringSplitOptions.RemoveEmptyEntries);
                for (var i = 0; i < tokens.Length; i += 2)
                {
                    if (!ValidateField(tokens[i], tokens[i + 1]))
                    {
                        valid = false;
                        break;
                    }
                    requiredFields.Remove(tokens[i]);
                }
                if (valid && requiredFields.Count == 0) ++solution;
            }
            return solution;
        }
    }
}
